src/structures/debruijn_graph.cpp

```diff
@@ -0,0 +1,203 @@
+    graph.reserveNode(static_cast<int>(reference.size() + 1 - len));
+    graph.reserveArc(static_cast<int>(reference.size() - len));
+
+    lemon::ListDigraph::Node prev_node = lemon::INVALID; // the previously processed node
+    for (size_t kmer : seqan3::views::kmer_hash(reference, seqan3::ungapped{len})) // iterate kmers
+    {
+        if (nodes(kmer) == lemon::INVALID) // the kmer is new
+        {
+            nodes.set(graph.addNode(), kmer); // add node
+            if (prev_node != lemon::INVALID)
+                arcs.set(graph.addArc(prev_node, nodes(kmer)), 0); // add arc
+
+            prev_node = nodes(kmer); // update processed node
+        }
+        else // abort: the kmer already exists (cycle)
+        {
+            len = 0;
+            return false;
+        }
+    }
+    return true;
+}
+
+void DeBruijnGraph::add_read(seqan3::dna4_vector const & read)
+{
+    assert(len > 0); // adding reads to an uninitialized graph is an error
+    lemon::ListDigraph::Node prev_node = lemon::INVALID; // the previously processed node
+    for (size_t kmer : seqan3::views::kmer_hash(read, seqan3::ungapped{len})) // iterate kmers
+    {
+        if (nodes(kmer) == lemon::INVALID) // kmer is not in graph: new node
+            nodes.set(graph.addNode(), kmer);
+
+        if (prev_node != lemon::INVALID)
+        {
+            lemon::ListDigraph::Arc arc = lemon::findArc(graph, prev_node, nodes(kmer));
+            if (arc != lemon::INVALID) // arc exists: increment read count
+                ++arcs[arc];
+            else
+                arcs.set(graph.addArc(prev_node, nodes(kmer)), 1); // new arc
+        }
+        prev_node = nodes(kmer); // update processed node
+    }
+    viable = std::nullopt; // we do not know if graph is viable anymore
+}
+
+bool DeBruijnGraph::is_viable()
+{
+    if (!viable) // status is unknown
+    {
+        if (len == 0 || !lemon::dag(graph)) // graph is uninitialized or has cycles
+        {
+            viable = false;
+        }
+        else // check if graph is complex enough (number of non-unique nodes < 4 * number of unique nodes)
+        {
+            int num_unique{}; // number of unique kmers (0-1 in and out arcs)
+            int num_non_unique{}; // number of non-unique kmers (>1 in or out arcs)
+            for (lemon::ListDigraph::NodeIt node(graph); node != lemon::INVALID; ++node)
+            {
+                lemon::ListDigraph::InArcIt in_arc(graph, node);
+                lemon::ListDigraph::OutArcIt out_arc(graph, node);
+                bool in = in_arc == lemon::INVALID || (arcs[in_arc] < 2 && ++in_arc == lemon::INVALID);
+                bool out = out_arc == lemon::INVALID || (arcs[out_arc] < 2 && ++out_arc == lemon::INVALID);
+                if (in && out)
+                    ++num_unique;
+                else
+                    ++num_non_unique;
+            }
+            viable = num_non_unique < 4 * num_unique;
+        }
+    }
+    return *viable;
```

Comment:
Is there a reason to use `*viable` over `viable.value()`?

Reply:
Yes, a good reason: `viable.value()` performs a check whether `viable` is set, and `*viable` does not. The check is unnecessary, since I am sure here that a value is set.